mar/lambda_function.py

```diff
@@ -0,0 +1,154 @@
+from __future__ import print_function
+
+import urllib
+import sys
+import os
+import json
+import subprocess
+
+class Context(object):
+    def __init__(self, logger, metrics_writter, request, response, system_info):
+        self.logger = logger
+        self.metrics_writter = metrics_writter
+        self.request = request
+        self.response = response
+        self.system_info = system_info
+
+
+class BaseResponse(object):
+    """BaseResponse if the object providing standard response attributes and operations"""
+    def __init__(self, response_body=""):
+        """Constructor, construct a response
+
+        Parameters
+        ----------
+        response_body: Optionally give the response a non-empty body
+        """
+        self.headers = []
+        self.status_code = 200
+        self.body = response_body
+
+    def get_dict(self):
+        """Get the response in dict form"""
+        return {"headers":dict(self.headers), "body":self.body, "statusCode":self.status_code}
+
+    def add_response_property(self, key, value):
+        """Add a pair of property to response
+
+        Parameters
+        ----------
+        key: string
+            Key of the property
+        value: string
+            Value of the property
+        """
+        self.headers.append((key, value))
+
+    def set_response_status(self, code, reason_string=None):
+        """Set the response status
+
+        Parameters
+        ----------
+        code: int
+            Status code of the response
+        reason_string: string
+            Optional non-default response reason string
+        """
+        self.status_code = code
+
+    def set_response_body(self, response_body):
+        """Set the response body
+
+        Parameters
+        ----------
+        response_body: object
+            The body of response
+        """
+        self.body = response_body
+
+
+def download_url(url, target, retries=2, base_retry_interval=0.01):
+    """Download url to a file.
+    Parameters
+    ----------
+    url: string
+        url to the file to download
+    target: string
+        the target local path of the downloaded file
+    retries: int
+        the max number of retries allowed for urlretrieve
+    """
+    assert retries >= 1, "Number of retries should be at least 1"
+
+    retry = 0
+    while retry < retries:
+        try:
+            urllib.urlretrieve(url, target)
+            return
+        except Exception as e:
+            retry += 1
+            if retry < retries: 
+                time.sleep(2 ** retries * base_retry_interval)
+            else:
+                raise e
+
+
+def lambda_handler(event, context):
```

Comment:
this function looks noisy, i think it can be simplified for readability, pseudocode below
```
def lambda_handler:
  #separate method
   getmodel_archive -> reads config, download, unzip, return path to MAR
   #separate method
   get_inference_handler-> reads MANIFEST gets the inference method.
    
   create response 
   create context

   execute handler

   return response
```

Reply:
Modified accordingly